radix-transactions/src/manifest/static_resource_movements/typed_invocation.rs

```diff
@@ -1,622 +1,1127 @@
-                            method_name: &str,
-                            args: &::radix_common::prelude::ManifestValue
-                        ) -> Result<Self, StaticResourceMovementsError> {
-                            Ok(match method_name {
-                                $(
-                                    $method_name => {
-                                        let encoded = ::radix_common::prelude::manifest_encode(args)
-                                            .map_err(StaticResourceMovementsError::NativeArgumentsEncodeError)?;
-                                        let decoded = ::radix_common::prelude::manifest_decode(&encoded)
-                                            .map_err(StaticResourceMovementsError::NativeArgumentsDecodeError)?;
-                                        Self::$method_ident(decoded)
-                                    }
-                                )*
-                                _ => return Err(StaticResourceMovementsError::UnknownNativeMethod {
-                                    package: $package_address,
-                                    blueprint: stringify!($blueprint_name).to_string(),
-                                    method: method_name.to_string(),
-                                })
+                impl [< $blueprint_ident BlueprintMethod >] {
+                    #![allow(unreachable_patterns, unused_variables)]
+                    pub fn decode_invocation(
+                        // Only provided for nice errors. Not needed in the decoding.
+                        module_id: ModuleId,
+                        method_name: &str,
+                        args: &ManifestValue
+                    ) -> Result<Self, TypedManifestNativeInvocationError> {
+                        match method_name {
+                            $(
+                                $method_name => decode_args(args)
+                                    .map(Self::$method_ident)
+                                    .map_err(|error| {
+                                        TypedManifestNativeInvocationError::FailedToDecodeMethodInvocation {
+                                            blueprint_id: ::radix_common::prelude::BlueprintId::new(&$package_address, $blueprint_name),
+                                            module_id,
+                                            method_name: method_name.to_owned(),
+                                            args: args.clone(),
+                                            error
+                                        }
+                                    }),
+                            )*
+                            // If we get here then it means that an invalid method was called. We
+                            // have all of the methods on all blueprints we have supported so this
+                            // should be an error.
+                            _ => Err(TypedManifestNativeInvocationError::InvokedFnDoesntExist {
```

Comment:
It's not a function. Can I suggest we spell it out in full as `InvokedMethodDoesNotExist` or `InvokedMethodNotFoundOnNativeBlueprint`

Reply:
Yeah I wanted to get a more generic name that would capture method & function and thought that `fn` would be appropriate. I also thought of `export` or `function` since we refer to both methods and functions in the blueprint definition as just functions.